File: harness/js/crossfuzz.ts
/**
 * cross_fuzz harness for JavaScript/TypeScript targets running under Bun.
 *
 * Usage:
 *   import { fuzz } from "../../harness/js/crossfuzz";
 *   fuzz((input: Uint8Array): Uint8Array => { ... });
 *
 * Shared memory is mapped via Bun.mmap(). Coverage is collected from the
 * Istanbul __coverage__ global if present (enabled by preloading instrument.ts):
 *   bun run --preload ../../harness/js/instrument.ts ./target.ts
 */

import { readSync, writeSync, openSync } from "node:fs";

// Shared memory layout — must match pkg/coverage/shmem.go
const OFF_INPUT_LEN   = 8;
const OFF_OUTPUT_LEN  = 12;
const OFF_STATUS      = 16;
const INPUT_OFFSET    = 64;
const INPUT_SIZE      = 1_048_576;
const OUTPUT_OFFSET   = 64 + 1_048_576;
const OUTPUT_SIZE     = 1_048_576;
const COVERAGE_OFFSET = 64 + 1_048_576 * 2;
const COVERAGE_SIZE   = 65_536;
const TOTAL_SHM_SIZE  = COVERAGE_OFFSET + COVERAGE_SIZE;
const MAX_OUTPUT      = 1_048_576;

const STATUS_OK    = 0;
const STATUS_ERROR = 1;

// ---- Settings ----

export interface Settings {
  instrument?: boolean;
  warmup?: number;
  transform?: boolean;
  hinting?: boolean;
}

function resolveSettings(s?: Settings): Required<Settings> {
  return {
    instrument: s?.instrument ?? true,
    warmup: s?.warmup ?? 0,
    transform: s?.transform ?? false,
    hinting: s?.hinting ?? false,
  };
}

// ---- Target types ----

export type FuzzFn = (input: Uint8Array) => Uint8Array | Buffer | Promise<Uint8Array> | Promise<Buffer>;

export interface FilterResult {
  output: Uint8Array;
  accepted: boolean;
}

export type FilterFn = (input: Uint8Array) => FilterResult | Promise<FilterResult>;

export type CompareFn = (input: Uint8Array, names: string[], outputs: Uint8Array[]) => string | Promise<string>;

// ---- Shared memory helpers ----

function readU32LE(arr: Uint8Array, off: number): number {
  return (arr[off] | arr[off + 1] << 8 | arr[off + 2] << 16 | arr[off + 3] << 24) >>> 0;
}

function writeU32LE(arr: Uint8Array, off: number, v: number): void {
  arr[off]     =  v         & 0xFF;
  arr[off + 1] = (v >>> 8)  & 0xFF;
  arr[off + 2] = (v >>> 16) & 0xFF;
  arr[off + 3] = (v >>> 24) & 0xFF;
}

// ---- Pipe protocol (fd 3 = cmd in, fd 4 = resp out) ----

function readMsg(): Uint8Array | null {
  const header = new Uint8Array(4);
  let total = 0;
  while (total < 4) {
    const n = readSync(3, header, total, 4 - total, null);
    if (n === 0) return null;
    total += n;
  }
  const len = (header[0] << 24 | header[1] << 16 | header[2] << 8 | header[3]) >>> 0;
  if (len === 0 || len > 1 << 20) return null;

  const payload = new Uint8Array(len);
  total = 0;
  while (total < len) {
    const n = readSync(3, payload, total, len - total, null);
    if (n === 0) return null;
    total += n;
  }
  return payload;
}

function writeMsg(json: string): void {
  const payload = Buffer.from(json, "utf8");
  const header = new Uint8Array(4);
  const plen = payload.length;
  header[0] = (plen >>> 24) & 0xFF;
  header[1] = (plen >>> 16) & 0xFF;
  header[2] = (plen >>> 8)  & 0xFF;
  header[3] =  plen         & 0xFF;
  writeSync(4, header);
  writeSync(4, payload);
}

// ---- Istanbul coverage ----

type IstanbulCovData = {
  s: Record<string, number>;
  f: Record<string, number>;
  b: Record<string, number[]>;
};

function hashSlot(fileIdx: number, kind: number, counterIdx: number): number {
  let h = (Math.imul(fileIdx, 0x9E3779B9) >>> 0);
  h = (Math.imul(h ^ kind,        0xBF58476D) >>> 0);
  h = (Math.imul(h ^ counterIdx,  0x94D049BB) >>> 0);
  return (h ^ (h >>> 16)) & 0xFFFF;
}

function resetIstanbulCounters(): void {
  const cov = (globalThis as any).__coverage__ as Record<string, IstanbulCovData> | undefined;
  if (!cov) return;
  for (const data of Object.values(cov)) {
    for (const k of Object.keys(data.s)) data.s[k] = 0;
    for (const k of Object.keys(data.f)) data.f[k] = 0;
    for (const k of Object.keys(data.b)) data.b[k] = data.b[k].map(() => 0);
  }
}

function collectCoverageInto(bitmap: Uint8Array): void {
  const cov = (globalThis as any).__coverage__ as Record<string, IstanbulCovData> | undefined;
  if (!cov) return;

  bitmap.fill(0);

  let fileIdx = 0;
  for (const data of Object.values(cov)) {
    const fi = fileIdx++;

    for (const [k, v] of Object.entries(data.s) as [string, number][]) {
      if (v > 0) {
        const idx = hashSlot(fi, 0, parseInt(k));
        const val = Math.min(v, 255);
        if (val > bitmap[idx]) bitmap[idx] = val;
      }
    }

    for (const [k, v] of Object.entries(data.f) as [string, number][]) {
      if (v > 0) {
        const idx = hashSlot(fi, 1, parseInt(k));
        const val = Math.min(v, 255);
        if (val > bitmap[idx]) bitmap[idx] = val;
      }
    }

    for (const [k, branches] of Object.entries(data.b) as [string, number[]][]) {
      for (let i = 0; i < branches.length; i++) {
        const v = branches[i];
        if (v > 0) {
          const idx = hashSlot(fi, 2 + i, parseInt(k));
          const val = Math.min(v, 255);
          if (val > bitmap[idx]) bitmap[idx] = val;
        }
      }
    }
  }
}

// ---- Standalone functions ----

export function openShm(): Uint8Array {
  const shmPath = process.env.CROSSFUZZ_SHM;
  if (!shmPath) {
    throw new Error("crossfuzz: CROSSFUZZ_SHM not set");
  }
  if (typeof (Bun as any).mmap !== "function") {
    throw new Error("crossfuzz: Bun.mmap not available — upgrade to Bun >= 1.0");
  }
  return (Bun as any).mmap(shmPath);
}

export function startInstrumentation(): void {
  // Istanbul instrumentation is started by the preload script; nothing to do here.
}

export function clearInstrumentation(): void {
  resetIstanbulCounters();
}

export function collectInstrumentation(bitmap: Uint8Array): void {
  collectCoverageInto(bitmap);
}

export function setStatus(shm: Uint8Array, status: number): void {
  writeU32LE(shm, OFF_STATUS, status);
}

// ---- Fuzz entry point ----

export async function fuzz(target: FuzzFn, settings?: Settings): Promise<void> {
  const s = resolveSettings(settings);

  const shmPath = process.env.CROSSFUZZ_SHM;
  if (!shmPath) {
    process.stderr.write("crossfuzz: CROSSFUZZ_SHM not set\n");
    process.exit(1);
  }
  if (typeof (Bun as any).mmap !== "function") {
    process.stderr.write("crossfuzz: Bun.mmap not available — upgrade to Bun >= 1.0\n");
    process.exit(1);
  }

  const shm: Uint8Array = (Bun as any).mmap(shmPath);

  writeMsg('{"type":"ready"}');

  for (;;) {
    const msgBytes = readMsg();
    if (!msgBytes) break;

    const msg = JSON.parse(Buffer.from(msgBytes).toString("utf8")) as { type: string };

    if (msg.type === "shutdown") break;

    if (msg.type === "fuzz") {
      const inputLen = readU32LE(shm, OFF_INPUT_LEN);
      const input = shm.slice(INPUT_OFFSET, INPUT_OFFSET + inputLen);

      resetIstanbulCounters();

      let output: Uint8Array | Buffer;
      let status = STATUS_OK;
      try {
        output = await target(input);
        if (!output) output = new Uint8Array(0);
      } catch {
        output = new Uint8Array(0);
        status = STATUS_ERROR;
      }

      const outLen = Math.min(output.length, MAX_OUTPUT);
      const outBytes =
        output instanceof Buffer
          ? new Uint8Array(output.buffer, output.byteOffset, outLen)
          : output.subarray(0, outLen);

      writeU32LE(shm, OFF_OUTPUT_LEN, outLen);
      writeU32LE(shm, OFF_STATUS, status);
      shm.set(outBytes, OUTPUT_OFFSET);

      if (s.instrument) {
        collectCoverageInto(shm.subarray(COVERAGE_OFFSET, COVERAGE_OFFSET + COVERAGE_SIZE));
      }

      writeMsg('{"type":"fuzz_result","ok":true}');
    }
  }
}

// ---- Filter entry point ----

export async function filter(target: FilterFn, settings?: Settings): Promise<void> {
  const s = resolveSettings(settings);

  const shmPath = process.env.CROSSFUZZ_SHM;
  if (!shmPath) {
    process.stderr.write("crossfuzz: CROSSFUZZ_SHM not set\n");
    process.exit(1);
  }
  if (typeof (Bun as any).mmap !== "function") {
    process.stderr.write("crossfuzz: Bun.mmap not available — upgrade to Bun >= 1.0\n");
    process.exit(1);
  }

  const shm: Uint8Array = (Bun as any).mmap(shmPath);

  writeMsg('{"type":"ready"}');

  for (;;) {
    const msgBytes = readMsg();
    if (!msgBytes) break;

    const msg = JSON.parse(Buffer.from(msgBytes).toString("utf8")) as { type: string };

    if (msg.type === "shutdown") break;

    if (msg.type === "filter") {
      const inputLen = readU32LE(shm, OFF_INPUT_LEN);
      const input = shm.slice(INPUT_OFFSET, INPUT_OFFSET + inputLen);

      let result: FilterResult;
      try {
        result = await target(input);
      } catch {
        result = { output: new Uint8Array(0), accepted: false };
      }

      if (result.accepted) {
        if (s.transform && result.output && result.output.length > 0) {
          const outLen = Math.min(result.output.length, MAX_OUTPUT);
          writeU32LE(shm, OFF_OUTPUT_LEN, outLen);
          shm.set(result.output.subarray(0, outLen), OUTPUT_OFFSET);
        } else {
          // Copy input to output region
          writeU32LE(shm, OFF_OUTPUT_LEN, inputLen);
          shm.set(input, OUTPUT_OFFSET);
        }
        writeMsg('{"type":"filter_result","ok":true}');
      } else {
        writeU32LE(shm, OFF_OUTPUT_LEN, 0);
        writeMsg('{"type":"filter_result","ok":false}');
      }
    }
  }
}

// ---- Compare entry point ----

export async function compare(target: CompareFn, settings?: Settings): Promise<void> {
  const _ = resolveSettings(settings);

  const targetsJSON = process.env.CROSSFUZZ_SHM_TARGETS;
  if (!targetsJSON) {
    process.stderr.write("crossfuzz: CROSSFUZZ_SHM_TARGETS not set\n");
    process.exit(1);
  }
  if (typeof (Bun as any).mmap !== "function") {
    process.stderr.write("crossfuzz: Bun.mmap not available — upgrade to Bun >= 1.0\n");
    process.exit(1);
  }

  const targetPaths: Record<string, string> = JSON.parse(targetsJSON);
  const targetShms: Record<string, Uint8Array> = {};
  for (const [name, path] of Object.entries(targetPaths)) {
    targetShms[name] = (Bun as any).mmap(path);
  }

  writeMsg('{"type":"ready"}');

  for (;;) {
    const msgBytes = readMsg();
    if (!msgBytes) break;

    const msg = JSON.parse(Buffer.from(msgBytes).toString("utf8")) as {
      type: string;
      targets?: string[];
    };

    if (msg.type === "shutdown") break;

    if (msg.type === "compare") {
      const reqTargets = msg.targets || [];
      const names: string[] = [];
      const outputs: Uint8Array[] = [];
      let input: Uint8Array | null = null;

      for (const name of reqTargets) {
        const tshm = targetShms[name];
        if (!tshm) continue;

        if (input === null) {
          const inLen = Math.min(readU32LE(tshm, OFF_INPUT_LEN), INPUT_SIZE);
          input = tshm.slice(INPUT_OFFSET, INPUT_OFFSET + inLen);
        }

        const outLen = Math.min(readU32LE(tshm, OFF_OUTPUT_LEN), OUTPUT_SIZE);
        outputs.push(tshm.slice(OUTPUT_OFFSET, OUTPUT_OFFSET + outLen));
        names.push(name);
      }

      if (!input) input = new Uint8Array(0);

      let mismatch: string;
      try {
        mismatch = await target(input, names, outputs);
      } catch (e: any) {
        mismatch = `compare exception: ${e?.message || e}`;
      }

      if (mismatch) {
        const escaped = mismatch.replace(/\\/g, "\\\\").replace(/"/g, '\\"').replace(/\n/g, "\\n");
        writeMsg(`{"type":"compare_result","error":"${escaped}"}`);
      } else {
        writeMsg('{"type":"compare_result"}');
      }
    }
  }
}

File: harness/python/crossfuzz.py
"""cross_fuzz Python harness.

Handles IPC between the cross_fuzz coordinator and a Python target. You write
the target function; the harness manages shared memory, pipes, and coverage.

Usage::

    import sys, os
    sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../../harness/python'))
    import crossfuzz

    def my_target(data: bytes) -> bytes:
        return data  # process input, return output

    crossfuzz.fuzz(my_target)

Coverage is collected via the `coverage` library (``pip install coverage``).
Branch arcs are hashed into the 64 KB shared-memory bitmap using the same
multiplicative hash scheme as the JS harness.
"""

import json
import mmap as _mmap
import os
import struct
import sys

# ── Shared memory layout (must match pkg/coverage/shmem.go) ──────────────────
_OFF_INPUT_LEN  = 8
_OFF_OUTPUT_LEN = 12
_OFF_STATUS     = 16
_INPUT_OFFSET   = 0x000040
_INPUT_SIZE     = 1 << 20       # 1 MB
_OUTPUT_OFFSET  = 0x100040
_OUTPUT_SIZE    = 1 << 20       # 1 MB
_COV_OFFSET     = 0x200040
_COV_SIZE       = 1 << 16       # 64 KB
_TOTAL_SIZE     = 0x210040

_STATUS_OK    = 0
_STATUS_ERROR = 1

_CMD_FD  = 3   # coordinator → harness (read)
_RESP_FD = 4   # harness → coordinator (write)

# ── Protocol ──────────────────────────────────────────────────────────────────

def _read_exact(fd, n):
    buf = bytearray(n)
    mv  = memoryview(buf)
    pos = 0
    while pos < n:
        chunk = os.read(fd, n - pos)
        if not chunk:
            return None
        mv[pos:pos + len(chunk)] = chunk
        pos += len(chunk)
    return bytes(buf)


def _read_msg(fd):
    hdr = _read_exact(fd, 4)
    if hdr is None:
        return None
    length = struct.unpack('>I', hdr)[0]
    if length == 0 or length > (1 << 20):
        return None
    return _read_exact(fd, length)


def _write_msg(fd, text):
    if isinstance(text, str):
        text = text.encode()
    os.write(fd, struct.pack('>I', len(text)) + text)


def _escape_json(s):
    return s.replace('\\', '\\\\').replace('"', '\\"').replace('\n', '\\n').replace('\r', '\\r')


# ── Shared memory ─────────────────────────────────────────────────────────────

class _SHM:
    """Read-write mmap of the target's shared memory region."""
    __slots__ = ('_mm',)

    def __init__(self, path):
        fd = os.open(path, os.O_RDWR)
        self._mm = _mmap.mmap(fd, _TOTAL_SIZE)
        os.close(fd)

    def read_input(self):
        n = min(struct.unpack_from('<I', self._mm, _OFF_INPUT_LEN)[0], _INPUT_SIZE)
        self._mm.seek(_INPUT_OFFSET)
        return self._mm.read(n)

    def write_output(self, data):
        n = min(len(data), _OUTPUT_SIZE)
        struct.pack_into('<I', self._mm, _OFF_OUTPUT_LEN, n)
        self._mm.seek(_OUTPUT_OFFSET)
        self._mm.write(data[:n])

    def set_output_len(self, n):
        struct.pack_into('<I', self._mm, _OFF_OUTPUT_LEN, n)

    def set_status(self, s):
        struct.pack_into('<I', self._mm, _OFF_STATUS, s)

    def cov_view(self):
        """Writable memoryview of the 64 KB coverage bitmap."""
        return memoryview(self._mm)[_COV_OFFSET:_COV_OFFSET + _COV_SIZE]


class _ROShm:
    """Read-only mmap of a target's shared memory (used by the compare harness)."""
    __slots__ = ('_mm',)

    def __init__(self, path):
        fd = os.open(path, os.O_RDONLY)
        self._mm = _mmap.mmap(fd, _TOTAL_SIZE, access=_mmap.ACCESS_READ)
        os.close(fd)

    def read_input(self):
        n = min(struct.unpack_from('<I', self._mm, _OFF_INPUT_LEN)[0], _INPUT_SIZE)
        self._mm.seek(_INPUT_OFFSET)
        return self._mm.read(n)

    def read_output(self):
        n = min(struct.unpack_from('<I', self._mm, _OFF_OUTPUT_LEN)[0], _OUTPUT_SIZE)
        self._mm.seek(_OUTPUT_OFFSET)
        return self._mm.read(n)


# ── Coverage ──────────────────────────────────────────────────────────────────

def _hash_slot(file_idx, from_line, to_line):
    """Map a coverage arc to a 16-bit bitmap index."""
    h = (file_idx * 0x9E3779B9) & 0xFFFFFFFF
    h = ((h ^ (from_line & 0xFFFFFFFF)) * 0xBF58476D) & 0xFFFFFFFF
    h = ((h ^ (to_line   & 0xFFFFFFFF)) * 0x94D049BB) & 0xFFFFFFFF
    return (h ^ (h >> 16)) & 0xFFFF


def _collect_into(cov_data, file_index, bv):
    """Accumulate coverage arc counts from cov_data into bytearray bv.

    file_index is a dict mapping absolute filename → stable integer index that
    grows monotonically across iterations so bitmap slots stay consistent.
    """
    for filename in cov_data.measured_files():
        fi = file_index.setdefault(filename, len(file_index))
        arcs = cov_data.arcs(filename)
        if not arcs:
            continue
        for from_line, to_line in arcs:
            slot = _hash_slot(fi, from_line, to_line)
            v = bv[slot]
            if v < 255:
                bv[slot] = v + 1


# ── Settings ──────────────────────────────────────────────────────────────────

class Settings:
    """Harness configuration. All fields have safe defaults."""

    def __init__(self, *, instrument=True, warmup=0, transform=False):
        # Enable branch-arc coverage collection via the `coverage` library.
        # Default: True. Set False when the harness is a thin HTTP client and
        # coverage comes from an instrumented server process.
        self.instrument = instrument

        # Run the target this many extra times on the first input before
        # entering the main loop (pre-heating the interpreter). Default: 0.
        self.warmup = warmup

        # Filter mode only: when True, the bytes returned by the filter replace
        # the original input for downstream targets. Default: False.
        self.transform = transform


# ── Entry points ──────────────────────────────────────────────────────────────

def fuzz(target, settings=None):
    """Enter the persistent fuzzing loop.

    ``target(data: bytes) -> bytes`` — raise an exception to signal an error.
    """
    if settings is None:
        settings = Settings()

    shm_path = os.environ.get('CROSSFUZZ_SHM')
    if not shm_path:
        sys.stderr.write('crossfuzz: CROSSFUZZ_SHM not set\n')
        sys.exit(1)
    shm = _SHM(shm_path)

    cov        = None
    file_index = {}   # filename → stable integer index
    warmed_up  = False

    if settings.instrument:
        try:
            import coverage as _cov_lib
        except ImportError:
            sys.stderr.write(
                'crossfuzz: coverage library not installed (pip install coverage)\n'
            )
            sys.exit(1)
        cov = _cov_lib.Coverage(
            branch=True,
            data_file=None,
            omit=[os.path.abspath(__file__)],
        )

    _write_msg(_RESP_FD, '{"type":"ready"}')

    while True:
        raw = _read_msg(_CMD_FD)
        if raw is None:
            return

        try:
            msg = json.loads(raw)
        except Exception:
            continue

        t = msg.get('type')

        if t == 'shutdown':
            return

        elif t == 'fuzz':
            inp = shm.read_input()

            # Warmup: pre-heat the interpreter on the first real input.
            if settings.warmup > 0 and not warmed_up:
                for _ in range(settings.warmup):
                    try:
                        target(inp)
                    except Exception:
                        pass
                warmed_up = True

            if cov is not None:
                cov.start()

            try:
                out = target(inp)
                if isinstance(out, str):
                    out = out.encode()
                shm.write_output(bytes(out))
                shm.set_status(_STATUS_OK)
                resp = '{"type":"fuzz_result","ok":true}'
            except Exception as e:
                shm.set_output_len(0)
                shm.set_status(_STATUS_ERROR)
                resp = f'{{"type":"fuzz_result","error":"{_escape_json(str(e))}"}}'

            if cov is not None:
                cov.stop()
                bv = bytearray(_COV_SIZE)
                _collect_into(cov.get_data(), file_index, bv)
                shm.cov_view()[:] = bv
                cov.erase()

            _write_msg(_RESP_FD, resp)

        elif t == 'ping':
            _write_msg(_RESP_FD, '{"type":"pong"}')


def filter(target, settings=None):
    """Enter the persistent filter loop.

    ``target(data: bytes) -> tuple[bytes, bool]`` — returns ``(output, accepted)``.
    """
    if settings is None:
        settings = Settings()

    shm_path = os.environ.get('CROSSFUZZ_SHM')
    if not shm_path:
        sys.stderr.write('crossfuzz: CROSSFUZZ_SHM not set\n')
        sys.exit(1)
    shm = _SHM(shm_path)

    _write_msg(_RESP_FD, '{"type":"ready"}')

    while True:
        raw = _read_msg(_CMD_FD)
        if raw is None:
            return

        try:
            msg = json.loads(raw)
        except Exception:
            continue

        t = msg.get('type')

        if t == 'shutdown':
            return

        elif t == 'filter':
            inp = shm.read_input()
            try:
                out, accepted = target(inp)
            except Exception:
                out, accepted = b'', False

            if accepted:
                if settings.transform and out:
                    shm.write_output(bytes(out))
                else:
                    shm.write_output(inp)
                _write_msg(_RESP_FD, '{"type":"filter_result","ok":true}')
            else:
                shm.set_output_len(0)
                _write_msg(_RESP_FD, '{"type":"filter_result"}')

        elif t == 'ping':
            _write_msg(_RESP_FD, '{"type":"pong"}')


def compare(target, settings=None):
    """Enter the persistent compare loop.

    ``target(input: bytes, names: list[str], outputs: list[bytes]) -> str``
    — returns ``''`` if all outputs agree, or a non-empty mismatch description.

    The compare process reads ``CROSSFUZZ_SHM_TARGETS`` (a JSON map of
    ``{"name": "/path/to/shm", ...}``) set by the coordinator.
    """
    targets_json = os.environ.get('CROSSFUZZ_SHM_TARGETS')
    if not targets_json:
        sys.stderr.write('crossfuzz: CROSSFUZZ_SHM_TARGETS not set\n')
        sys.exit(1)

    try:
        target_paths = json.loads(targets_json)
    except Exception as e:
        sys.stderr.write(f'crossfuzz: parse CROSSFUZZ_SHM_TARGETS: {e}\n')
        sys.exit(1)

    target_shms = {}
    for name, path in target_paths.items():
        try:
            target_shms[name] = _ROShm(path)
        except OSError as e:
            sys.stderr.write(f'crossfuzz: open target SHM {name} ({path}): {e}\n')
            sys.exit(1)

    _write_msg(_RESP_FD, '{"type":"ready"}')

    while True:
        raw = _read_msg(_CMD_FD)
        if raw is None:
            return

        try:
            msg = json.loads(raw)
        except Exception:
            continue

        t = msg.get('type')

        if t == 'shutdown':
            return

        elif t == 'compare':
            names   = msg.get('targets', [])
            inp     = None
            outputs = []

            for name in names:
                ts = target_shms.get(name)
                if ts is None:
                    outputs.append(b'')
                    continue
                if inp is None:
                    inp = ts.read_input()
                outputs.append(ts.read_output())

            if inp is None:
                inp = b''

            try:
                mismatch = target(inp, names, outputs)
            except Exception as e:
                mismatch = f'compare exception: {e}'

            if mismatch:
                _write_msg(
                    _RESP_FD,
                    f'{{"type":"compare_result","error":"{_escape_json(str(mismatch))}"}}',
                )
            else:
                _write_msg(_RESP_FD, '{"type":"compare_result"}')

        elif t == 'ping':
            _write_msg(_RESP_FD, '{"type":"pong"}')
